### PR TITLE
[SPARK-34037][SQL] Remove unnecessary upcasting for Avg & Sum which handle by themself internally

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?
The type-coercion for numeric types of average and sum is not necessary at all, as the resultType and sumType can prevent the overflow.

### Why are the changes needed?

rm unnecessary logic which may cause potential performance regressions

### Does this PR introduce _any_ user-facing change?

no
### How was this patch tested?

tpcds tests for plan

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
cc @maropu and @cloud-fan

**Comment by viirya** [comment]
Hmm? Can you elaborate the issue more clearly? I don't get what going to be wrong with the current PR description. What the bug is, and how it affects query result?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 1]
```diff
@@ -634,17 +634,6 @@ object TypeCoercion {
 
         m.copy(newKeys.zip(newValues).flatMap { case (k, v) => Seq(k, v) })
 
-      // Promote SUM, SUM DISTINCT and AVERAGE to largest types to prevent overflows.
-      case s @ Sum(e @ DecimalType()) => s // Decimal is already the biggest.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:638]: There is a reason about promoting these aggregation functions, why removing them?

**Comment by cloud-fan** [comment]
Let's add a UT to show the bug directly. TPCDS query plans are hard to read.

**Comment by yaooqinn** [comment]
Hi @cloud-fan @viirya I added both new UT and e2e tests to secure this fix, also the PR desc has been updated too, thanks.

**Comment by yaooqinn** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:638]: the type-coercion for numeric types of average and sum is not necessary at all, as the resultType and sumType can prevent the overflow. and it causes the issue here

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:638]: Yea, `Sum`/`Average` already casts the inputs internally.

**Comment by yaooqinn** [comment]
retest this please

**Comment by cloud-fan** [comment]
@yaooqinn can we update the PR title and description? Now it's a simple improvement to avoid unnecessary casts for sum/avg.

**Comment by yaooqinn** [comment]
updated thanks

**Review by cloud-fan — APPROVED** [review]

**Review by viirya — APPROVED** [review]
looks making sense.

**Comment by viirya** [comment]
@yaooqinn Can you also update the JIRA accordingly as the current PR title and description?

**Comment by viirya** [comment]
Thanks! Merging to master.

**Pull request closed without merging**